test/jdk/java/nio/channels/DatagramChannel/Connect.java

```diff
@@ -48,84 +50,78 @@ static void test() throws Exception {
         invoke(a, r);
     }
 
-    static void invoke(Sprintable reader, Sprintable writer) throws Exception {
-
-        Thread writerThread = new Thread(writer);
-        writerThread.start();
-
-        Thread readerThread = new Thread(reader);
-        readerThread.start();
-
-        writerThread.join();
-        readerThread.join();
-
-        reader.throwException();
-        writer.throwException();
+    static void invoke(Runnable reader, Runnable writer) throws CompletionException {
+        CompletableFuture<Void> f1 = CompletableFuture.runAsync(writer);
+        CompletableFuture<Void> f2 = CompletableFuture.runAsync(reader);
+        wait(f1, f2);
     }
 
-    public interface Sprintable extends Runnable {
-        public void throwException() throws Exception;
+    // This method waits until one of the given CompletableFutures completes exceptionally. In which case, it stops waiting for the other futures and
+    // throws a CompletionException. Otherwise, will wait for all futures to complete successfully.
+    private static void wait(CompletableFuture<?>... futures) throws CompletionException {
+        CompletableFuture<?> future = CompletableFuture.allOf(futures);
+        Stream.of(futures)
+                .forEach(f -> {
+                    f.exceptionally(ex -> {
+                        future.completeExceptionally(ex);
+                        return null;
+                    });
+                });
+        future.join();
```

Comment:
Would ExecutorService invokeAny help you hear?

Reply:
I don't think so - we kind of want to do the opposite: what we want is stop the test if any of the two tasks throw an exception. The previous version of the test might have been blocked in the first call to `Thread::join` and not even noticed that the other thread had exited - which meant that the test would fail in timeout and not even report the exception that made the other thread terminate. Here the allOf `future` will be completed as soon as: one task throws an exception (thanks to the dependent action), or the two tasks complete.
There are many ways to do that - but I find the solution that Conor came up with rather elegant.

Reply:
Okay, they might new combinators here in the future that makes this a bit easier.